### PR TITLE
Big 'ol refactor

### PR DESCRIPTION
**Pull request by jocr1627**
I implemented a whole lot of refactors in hopes of solving future problems. Here are the major changes:

* Actions are entities.
  * We had talked about this and I believe it will be especially useful in a game like DnD where actions themselves have state that can be manipulated prior to their execution.
  * As a part of this I removed the concept of passing in a list of "reactions" since any child entities which are some form of listener should be able to react once they are in the same entity tree.
  * I updated the entity class to maintain a concept of root, listeners, and a complete descendants registry as children are added and removed. The idea here is I can always inspect my entity's root to look up other entities O(1) in the same tree by id. If a child is removed from an entity tree, it takes all its downstream entities with it and becomes the new root. This should allow you to flexibly build up and mutate an entity/action tree even if it is not yet a subtree of a Game, add it in when appropriate, remove it as needed, and the game will always maintain an accurate registry.
* Diffs are automatic now.
  * I made it so your root entity will just keep track of diffs for each scope for you using a start_diff and end_diff method. Now when you call set within an execute method the diff will be updated without having to merge and return it.
* State has better performance.
  * I added an inspect/inspectIn method to state which helps with performance in certain circumstances. A good example of this is when you want to know the length of a list in state without deepcopying the whole list.
* The root has a class registry.
  * The entity_classes map is horrible thing that I hate that I needed in order to get past python's horrible import system. I haven't found a better way to have actions pass around classes (e.g. I want to tell an action downstream what class to load and resolve during its execute- see request.py). If you try to import the class in the downstream file you'll get a cyclical dependency. If you keep it in state then you'll end up deepcopying an entire class. If you don't keep it in state then there isn't a mechanism for other actions to mutate it and benefit from the whole diff/listener structure.

### CODE REVIEW

File: engine/entity.py [hunk 1]
```diff
@@ -1,81 +1,131 @@
...
+
+      for descendant_id in child.descendants:
+        descendant = child.descendants[descendant_id]
+        self.root.descendants[descendant_id] = descendant
+        descendant.root = self.root
+
+      for listener_id in child.listeners:
+        self.root.listeners[listener_id] = child.listeners[listener_id]
+
+      child.descendants = {}
```
Comment (nigelrobbins3) [inline engine/entity.py]: Why do we blow away child descendants here?

File: examples/go_fish/go_fish.py [hunk 2]
```diff
@@ -1,10 +1,18 @@
+import os
```
Comment (nigelrobbins3) [inline examples/go_fish/go_fish.py]: Unneeded?

**Review by nigelrobbins3 — APPROVED** [review]
Looks great. Was the performance really bad enough to notice with only a small deck of cards?

File: examples/go_fish/go_fish.py (hunk 2 above)
Comment (jocr1627) [reply examples/go_fish/go_fish.py]: yup- thanks

File: engine/entity.py (hunk 1 above)
Comment (jocr1627) [reply engine/entity.py]: @nigelrobbins3 in order to minimize unnecessary storage. The current intent is that you only need the root entity to keep the descendant registry. I added a hydrate method that you should be able to use instead of having to access the descendants registry directly, and it uses root.

**Comment by jocr1627** [comment]
Performance wasn't "bad" but it was notably worse and I wanted to make sure that we had the tools needed to not have deepcopy bite us in the ass. I also tested it with a couple orders of magnitude greater cards and players and profiled it to make sure none of the functions were particularly bad, and those checks were the top offenders.

File: engine/entity.py (hunk 1 above)
Comment (nigelrobbins3) [reply engine/entity.py]: :+1:

**Comment by nigelrobbins3** [comment]
Yeah, I can see that deepcopying the deck just to check if `get_should_react` is true every time could get expensive.
